### ISSUE TITLE
Cannot find the declaration of element 'nifty'

### ISSUE THREAD

**Issue by Cafnio** [nifty-gui/nifty-gui#335]
Hi there,

I wanted to start using Nifty-GUI in my jmonkey 3.0 project, though I cant get it to work.

Right now I have the following code:

<?xml version="1.0" encoding="UTF-8"?>
<nifty
        xmlns="http://nifty-gui.lessvoid.com/nifty-gui"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://nifty-gui.lessvoid.com/nifty-gui">
    <screen id="start" controller="mygame.Main">
        
    </screen>
</nifty>

When I put it on Design, it says:

Line -1 col :-1 : cvc-elt.1: Cannot find the declaration of element 'nifty'.,verticalAlignment=CENTER,verticalTextPosition=CENTER]

How can I fix it?

Thanks

**Comment by relu91** [comment]
Hello! It seems that you have syntax errors in your xml or is there some problem in github formatting? By the way if you want to try an other way to edit your nifty gui visually try my editor here: http://niftyeditor.it/ . I hope you enjoy it! :smile:

**Comment by void256** [comment]
Yes, Cris is right. You probably have a syntax error in your XML file that makes it not valid to the nifty.xsd. Can you provide the complete XML file for us to check?

Another cause might be your editor/ide that doesn't know about the nifty.xsd. In that case you'll need to find the nifty.xsd from the nifty.jar of the Nifty version you're using and tell your editor/ide to use it to validate your nifty xml file. Otherwise you might get the same error which basically tells you that it doesn't know about the nifty type and therefore it appears to be invalid.

Unfortunately the namespace URL used to be a link to a webpage on the old nifty-gui website that explains all of that XSD validation but after the website has been hacked I've not yet repaired/ported that part to the new website :disappointed:

**Comment by simplerun** [comment]
Same issue as @Cafnio  but I tried two xsd versions:

```
1. <nifty xmlns="http://nifty-gui.lessvoid.com/nifty-gui" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://nifty-gui.lessvoid.com/nifty-gui http://nifty-gui.lessvoid.com/nifty-gui">
```

```
2. <nifty xmlns="http://nifty-gui.sourceforge.net/nifty-1.3.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://nifty-gui.sourceforge.net/nifty-1.3.xsd http://nifty-gui.sourceforge.net/nifty-1.3.xsd">
```

I used generated files in jme3 and issue pop-up in SDK editor "Cannot find the declaration of element 'nifty' ". Also retriver output in jme sdk report that retrived location nifty-gui.sourceforge.net not found. On app run in jme sdk, app show nifty start screen but every action (mouse click) end up with error report in output window.
- jme as jMonkeyEngine

In packed jar nifty.jar in jme libs directory I found nifty-1.3.xsd but adding that jar into project crashed jme after app run in jme sdk.
I dont know how xsd schemas works. Just use them in project.

also downloaded nifty editor 0.59 to examine generated xml but there is no xsd declaration line.

Here is my file generated in jme SDK and partialy copied from jme source code pages. Generated GUI is as expected. But with errors about xsd and NiftyMethodInvoker.

```
<?xml version="1.0" encoding="UTF-8"?>
<!--<nifty xmlns="http://nifty-gui.lessvoid.com/nifty-gui" 
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
       xsi:schemaLocation="http://nifty-gui.lessvoid.com/nifty-gui http://nifty-gui.lessvoid.com/nifty-gui">-->
 <nifty xmlns="http://nifty-gui.sourceforge.net/nifty-1.3.xsd"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
        xsi:schemaLocation="http://nifty-gui.sourceforge.net/nifty-1.3.xsd http://nifty-gui.sourceforge.net/nifty-1.3.xsd">

    <!-- +++++++++++++++++++++++++++++++++++++++ -->
    <!-- start screen -->
    <!-- +++++++++++++++++++++++++++++++++++++++ -->
    <useStyles filename="nifty-default-styles.xml" />
  <useControls filename="nifty-default-controls.xml" />
  <screen id="start" controller="mygame.ScreenState" >
    <layer id="background" childLayout="center">
      <image filename="Interface/start-background.png"></image>
    </layer>
    <layer id="foreground" childLayout="vertical">
      <panel id="panel_top" height="25%" width="75%" align="center" childLayout="center">
        <text text="My Cool Game" font="Interface/Fonts/Default.fnt" width="100%" height="100%" />
      </panel>
      <panel id="panel_mid" height="50%" width="75%" align="center" childLayout="center">
        <text text="Here goes some text describing the game and the rules and stuff. Incidentally, the text is quite long and needs to wrap at the end of lines. ..." font="Interface/Fonts/Default.fnt" width="100%" height="100%" wrap="true" />
      </panel>
      <panel id="panel_bottom" height="25%" width="75%" align="center" childLayout="horizontal" >
        <panel id="panel_bottom_left" height="50%" width="50%" valign="center" childLayout="center">
          <control name="button" label="Start" id="StartButton" align="center" valign="center" visibleToMouse="true" >

              <interact onClick="startGame(hud)"/>
          </control>
        </panel>
        <panel id="panel_bottom_right" height="50%" width="50%" valign="center" childLayout="center">
          <control name="button" label="Quit" id="QuitButton" align="center" valign="center" visibleToMouse="true" >

              <interact onClick="quitGame()"/>
          </control>
        </panel>
      </panel>
    </layer>
  </screen>
  <screen id="hud" controller="mygame.ScreenState">
    <layer id="background" childLayout="center">
      <image filename="Interface/hud-frame.png"></image>
    </layer>
    <layer id="foreground" childLayout="horizontal">
      <panel id="panel_left" width="80%" height="100%" childLayout="vertical" ></panel>
      <panel id="panel_right" width="20%" height="100%" childLayout="vertical">
        <panel id="panel_top_right1" width="100%" height="15%" childLayout="center">
          <control name="label" color="#000" text="123" width="100%" height="100%" />
        </panel>
        <panel id="panel_top_right2" width="100%" height="15%" childLayout="center">
          <image filename="Interface/face1.png" valign="center" align="center" height="50%" width="30%" ></image>
        </panel>
        <panel id="panel_bot_right" width="100%" height="70%" valign="center" ></panel>
      </panel>
    </layer>
  </screen>
</nifty>
```

How can I get proper nifty-gui xsd ?

**Comment by simplerun** [comment]
Found alternative/temporary solution!
Unpacked nifty-gui.jar and copy nifty-1.3.xsd in my local apache. I used local IP 192.168.88.35 and then pulled out  xsd with (my working example):

```
<nifty xmlns="http://192.168.88.35/nifty-gui/nifty-1.3.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://192.168.88.35/nifty-gui/nifty-1.3.xsd http://192.168.88.35/nifty-gui/nifty-1.3.xsd">
```

Now I can see downloaded xsd in JME project directory under project/nbproject/private/retriver/192.168.88.35/nifty-gui and no more error "Cannot find the declaration of element 'nifty' "

Under same project for url http://nifty-gui.lessvoid.com/nifty-gui there is file with content:

```
<html>
<head>
<title>Nifty GUI - Blog moved to a new site</title>
<meta http-equiv="refresh" content="0; url=http://void256.github.io/nifty-gui/" />
</head>
<body>
<center>
    The Nifty GUI Blog moved to a new site:<br>
<a href="http://void256.github.io/nifty-gui/">http://void256.github.io/nifty-gui/</a>
</center>
</body>
</html>
```

and that last location return 404 code. Which means no so needed xsd.

This solution is temporary so my work may continue.

Waiting for any other solutions.

**Comment by relu91** [comment]
You found a clever solution :smile: . I think that you also point out that what problem is, the xsd is no more online. I don't know why it cause the exception in the editor tough. Are you using JME 3.1 alpha? If it's not you are using the old integrated editor. About my nifty editor try using the 0.6 version (it's exactly the same as the one integrated in JME 3.1 alpha ). I've just tried it and it's seems to not complain about the xsd.  

P.S. the 0.6 version is available in my githup repo.

**Comment by simplerun** [comment]
I use JME 3.0, wanted stable SDK for fast introduction into game programming. Almost everything  about JME on their pages is for 3.0 version. Have two PCs, one ubuntu 12 LTS, another ubuntu 14 LTS. Same errors on both.  Your point taken, must go forward although JME  3.1 is in alpha. 
... Downloaded NIfty editor 0.6

Thanks

**Comment by relu91** [comment]
You could stay with 3.0 if you will use edit the gui using the stand alone editor. It has a pretty decent integration with IDE (actually with whatever ide you are using). It's not well documented unfortunately :disappointed: sorry for that. Basically the steps too produce a working gui is:
1. Create a new gui file in JME IDE.
2. Open Nifty-Editor and open the file ( File-->Open )
3. Set Asset Folder in Nifty-Editor to JME asset folder. Go to Edit-->Asset Folder-->Set the jme asset folder of your project
4. Edit your gui as you want. Add images controls etc.. ( you can also add style and custom controls via File menu )
5. Then use the gui in your Game :smile:

**Comment by simplerun** [comment]
Everything works now. 1. xsd file is on rich localy, 2. Use JME3.0
Also use Nifty-editor as You instructed. Xml opens in NE0.6 but some tags missing after opening file.
Used xml that I posted few posts above. 
What is missing ?
 in panel id="panel_top" and in panel id="panel_mid" text tag missing, and tag

```
<interact onClick="startGame(hud)"/>
```

is missing.
If I do nothing some tags are missing from original xml.

Suggestions ?

**Comment by relu91** [comment]
Unfortunately that's a bug or better it's a not implemented feature :smile: . The editor should only skips that elements and save them after, but now it doesn't sorry. :cry:

**Comment by bgroenks96** [comment]
So @void256 is the hosting location for the XSD no longer valid?

**Comment by mkaring** [comment]
Isn't the XSD embedded into one of the jar files?

**Comment by void256** [comment]
I've recovered the following text from one of the old website backups. This text was available under http://nifty-gui.lessvoid.com/nifty-gui and is now available there as a plain-text file again.

Here is the text:

```
Hi there, seems like you're looking for the Nifty GUI XML-schema (XSD) file.
Well, the good news is, you've come to the kinda right place ;)

Starting with Nifty 1.4 the correct XSD will always be included in your nifty.jar.
To access it, you can simply unjar it (jar -xf nifty.jar) and find it in the file "nifty.xsd".
The reason for this and why the XSD is not directly available for download is to
allow us to make changes to the XSD more easily.

The latest version of the XSD is always availabe at github of course. So if you just
want to take a look you can do so here (Nifty 1.4 XSD at github):

https://github.com/void256/nifty-gui/blob/1.4/nifty-core/src/main/resources/nifty.xsd 

But make sure that the XSD matches the Nifty version you use. It's probably a good
idea to just use the "nifty.xsd" that ships with your Nifty version to be sure.
You know, the one you can find in the "nifty.jar".

Best regards,
void
```

In short, a XML namespace is usually a URI but it is not required by XML-Schema that an actual file exists at that URI (although sometimes it is).

The correct and current namespace for 1.4.x Nifty is `http://nifty-gui.lessvoid.com/nifty-gui`. All the other namespaces are old and not supported anymore (or require the corresponding old Nifty version).

JME probably shipped Nifty 1.3 for a long time which would explain the old namespace in the examples there or the code that was generated by the SDK. Current JME should come with Nifty 1.4 but I've not verified it right now.

I'll close this issue now. Feel free to re-open it if necessary!

**Comment by void256** [comment]
Oh and btw: http://nifty-gui.lessvoid.com/ now forwards to the correct URL as well: http://void256.github.io/blog/
